src/object.js

```diff
@@ -57,22 +58,24 @@ export const unwind = _.curry((prop, x) =>
 // this one's _actually_ just like mongo's `$unwind`
 export const unwindArray = _.curry((prop, xs) => _.flatMap(unwind(prop))(xs))
 
-export const isFlatObject = overNone([_.isPlainObject, _.isArray])
-
-// { a: { b: { c: 1 } } } => { 'a.b.c' : 1 }
-export const flattenObject = (input, paths) =>
-  reduceIndexed(
-    (output, value, key) =>
-      _.merge(
-        output,
-        (isFlatObject(value) ? singleObjectR : flattenObject)(
-          value,
-          dotJoinWith(isNotNil)([paths, key])
-        )
-      ),
+export const flattenObjectBy = _.curryN(2, (fn, value, rootPath) => {
+  let flat = fn(value, rootPath)
+  if (!_.isUndefined(flat)) return flat
+  if (!isTraversable(value)) return value
+  return reduceIndexed(
+    (result, x, k) =>
+      _.flow(
+        unless(_.isPlainObject, singleObject(k)),
```

Comment:
we shouldn't hard code this to `_.isPlainObject`

Reply:
in general, I don't follow why we couldn't just factor the `isFlatObject` method out of the existing implementation

Reply:
If we factor `isFlatObject`out, we're still assumming it'll return an object, which might not be the case. E.g. `_.merge(output, ...assumeToBeObj)`